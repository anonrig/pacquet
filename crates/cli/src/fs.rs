use std::{io, os, path::Path};

#[cfg(unix)]
pub fn symlink_dir<P: AsRef<Path>>(original: P, link: P) -> io::Result<()> {
    os::unix::fs::symlink(original, link)
}

#[cfg(windows)]
pub fn symlink_dir<P: AsRef<Path>>(original: P, link: P) -> io::Result<()> {
    os::windows::fs::symlink_dir(original, link)
}

#[cfg(test)]
pub fn get_filenames_in_folder(path: &Path) -> Vec<String> {
    let mut files = std::fs::read_dir(path)
        .unwrap()
        .map(|e| e.unwrap().file_name().to_string_lossy().to_string())
        .collect::<Vec<_>>();

    files.sort();
    files
}

#[cfg(test)]
pub fn get_all_folders(root: &std::path::PathBuf) -> Vec<String> {
    let mut files = Vec::new();
    for entry in walkdir::WalkDir::new(root) {
        let entry = entry.unwrap();
        let entry_path = entry.path().to_path_buf();
        if entry.file_type().is_dir() || entry.file_type().is_symlink() {
            // We need this mutation to ensure that both Unix and Windows paths resolves the same.
            // TODO: Find a better way to do this?
            let simple_path = entry_path
                .strip_prefix(root)
                .unwrap()
                .components()
                .map(|c| c.as_os_str().to_string_lossy().to_string())
                .collect::<Vec<_>>()
                .join("/");

            if !simple_path.is_empty() {
                files.push(simple_path.as_os_string());
            }
        }
    }
    files.sort();
    files
}


