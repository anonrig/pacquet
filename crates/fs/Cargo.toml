[package]
name                 = "pacquet-fs"
description          = "Filesystem utility functions used by pacquet"
version              = "0.0.1"
publish              = false
authors.workspace    = true
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
repository.workspace = true

[dependencies]
pacquet-task-queue = { workspace = true }

derive_more = { workspace = true }
miette      = { workspace = true }

[target.'cfg(windows)'.dependencies]
junction = { workspace = true }
