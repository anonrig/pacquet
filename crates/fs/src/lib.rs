mod ensure_file;
mod symlink_dir;

pub use ensure_file::*;
pub use symlink_dir::*;

pub mod file_mode;
