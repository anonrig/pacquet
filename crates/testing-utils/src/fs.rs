use std::{io, path::Path};

pub fn get_filenames_in_folder(path: &Path) -> Vec<String> {
    let mut files = std::fs::read_dir(path)
        .unwrap()
        .map(|e| e.unwrap().file_name().to_string_lossy().to_string())
        .collect::<Vec<_>>();

    files.sort();
    files
}

pub fn get_all_folders(root: &std::path::Path) -> Vec<String> {
    let mut files = Vec::new();
    for entry in walkdir::WalkDir::new(root) {
        let entry = entry.unwrap();
        let entry_path = entry.path();
        if entry.file_type().is_dir() || entry.file_type().is_symlink() {
            // We need this mutation to ensure that both Unix and Windows paths resolves the same.
            // TODO: Find a better way to do this?
            let simple_path = entry_path
                .strip_prefix(root)
                .unwrap()
                .components()
                .map(|c| c.as_os_str().to_str().expect("invalid UTF-8"))
                .collect::<Vec<_>>()
                .join("/");

            if !simple_path.is_empty() {
                files.push(simple_path);
            }
        }
    }
    files.sort();
    files
}

// Helper function to check if a path is a symlink or junction
pub fn is_symlink_or_junction(path: &Path) -> io::Result<bool> {
    #[cfg(windows)]
    return junction::exists(&path);

    #[cfg(not(windows))]
    return Ok(path.is_symlink());
}
