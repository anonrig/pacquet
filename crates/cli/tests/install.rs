use assert_cmd::prelude::*;
use command_extra::CommandExtra;
use pacquet_testing_utils::{
    bin::pacquet_with_temp_cwd,
    fs::{get_all_files, get_all_folders, is_symlink_or_junction},
};
use pipe_trait::Pipe;
use std::fs;

#[test]
fn should_install_dependencies() {
    let (command, root, workspace) = pacquet_with_temp_cwd(true);

    eprintln!("Creating package.json...");
    let manifest_path = workspace.join("package.json");
    let package_json_content = serde_json::json!({
        "dependencies": {
            "is-odd": "3.0.1",
        },
        "devDependencies": {
            "fast-decode-uri-component": "1.0.1",
        },
    });
    fs::write(&manifest_path, package_json_content.to_string()).expect("write to package.json");

    eprintln!("Executing command...");
    command.with_arg("install").assert().success();

    eprintln!("Make sure the package is installed");
    assert!(is_symlink_or_junction(&workspace.join("node_modules/is-odd")).unwrap());
    assert!(workspace.join("node_modules/.pacquet/is-odd@3.0.1").exists());

    eprintln!("Make sure it installs direct dependencies");
    assert!(!workspace.join("node_modules/is-number").exists());
    assert!(workspace.join("node_modules/.pacquet/is-number@6.0.0").exists());

    eprintln!("Make sure we install dev-dependencies as well");
    assert!(
        is_symlink_or_junction(&workspace.join("node_modules/fast-decode-uri-component")).unwrap()
    );
    assert!(workspace.join("node_modules/.pacquet/fast-decode-uri-component@1.0.1").is_dir());

    eprintln!("Snapshot");
    let workspace_folders = get_all_folders(&workspace);
    let store_files = get_all_files(&root.path().join("pacquet-store"));
    insta::assert_debug_snapshot!((workspace_folders, store_files));

    drop(root); // cleanup
}

#[test]
fn should_install_exec_files() {
    let (command, root, workspace) = pacquet_with_temp_cwd(true);

    eprintln!("Creating package.json...");
    let manifest_path = workspace.join("package.json");
    let package_json_content = serde_json::json!({
        "dependencies": {
            "pretty-exec": "0.3.10",
        },
    });
    fs::write(&manifest_path, package_json_content.to_string()).expect("write to package.json");

    eprintln!("Executing command...");
    command.with_arg("install").assert().success();

    eprintln!("Listing all files in the store...");
    let store_files = root.path().join("pacquet-store").pipe_as_ref(get_all_files);
    let (exec_files, non_exec_files): (Vec<_>, Vec<_>) =
        store_files.iter().partition(|path| path.ends_with("-exec"));

    eprintln!("Snapshot");
    insta::assert_debug_snapshot!(exec_files);

    #[cfg(unix)]
    {
        use pacquet_testing_utils::fs::is_path_executable;
        use pretty_assertions::assert_eq;

        eprintln!("All files that end with '-exec' are executable, others not");
        assert_eq!(
            store_files.iter().partition(|name| {
                root.path().join("pacquet-store").join(name).pipe_as_ref(is_path_executable)
            }),
            (exec_files, non_exec_files),
        );
    }

    #[cfg(windows)]
    let _ = non_exec_files;

    drop(root); // cleanup
}
