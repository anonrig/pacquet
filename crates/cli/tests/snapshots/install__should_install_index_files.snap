---
source: crates/cli/tests/install.rs
assertion_line: 134
expression: index_file_contents
---
v3/files/09/0a6758fac3c263f5f923075d986d2ed26ff74ca2c957e5b86b17e62342564ae142d5374d01ec5163c6f7091d93d2e0779c8da4083d8d0128f7408169781630-index.json:
  LICENSE:
    integrity: sha512-Vdy8f/8awcR1X0y/JLoh4NyuBiyALT25N9K2GC9FVN2F0UUpoMlojqyCWjRbY6uLLl12meK+BxeN9buKByI16w==
    mode: 420
    size: 1091
  README.md:
    integrity: sha512-QP4YssHyQQdbcycnEQMj8+KfnWyHZf8CIrzPo0eEyqirzRj/vUu/q5eeVHis4nOnTtOgoi8CYTNPp8WeAy7b3g==
    mode: 420
    size: 3495
  index.js:
    integrity: sha512-iB8QHS8YhvxfmsXI0l5VQJCOFAAxFp6kGkqZQc+V/icRwSksGZVgnG8XSzgVzA3MbNKhAzmjSNOf51vHe1O6ng==
    mode: 420
    size: 543
  package.json:
    integrity: sha512-VHvNaD4vn3sMJ5DGN7kjFTmUe+kjR5U0WS2RiGpW1iYv3B2M7SSsLMOt/aE5/jmdx1AtOCx6/FNuMFenzxHmwA==
    mode: 420
    size: 1381
v3/files/58/a80a5a0e5e531bd1646c16f05bdf6da1fb0174a1d9c2fede3e5f306cd4302c56049ddd5776bb5b3f32d9ffee4347b707a5a6a1d0f7a12e788d343d1d54c822-index.json:
  ".travis.yml":
    integrity: sha512-N46xbCMPzjVO9+cnIu5wjL7+qaSBqll7Bv78EvZlq2AXbL8E1rwVm+QszbxRdk1xUW+GuyigGw5gx2onU/HPRA==
    mode: 420
    size: 132
  LICENSE:
    integrity: sha512-5cJRTSsNPWvtrST911xV54AhHPBpLCH0ASAcLX36mQrQAYi9aLMMgmqqKnLsrRKnS8A0MDhPkzsXUQNd0bGMog==
    mode: 420
    size: 1174
  README.md:
    integrity: sha512-ymXfFutFCrY/kPEHqMyayC2pY1QI6BOm9tPhPPuniRezG2ZlJHnrulS2ElZmALKrG8FrgtCCxfh4A5Yajp45MQ==
    mode: 420
    size: 1941
  bench.js:
    integrity: sha512-YKtUhJ1s/XsFB2w7xNm6/+H4APTV4tZ8OcUhp6PZOzR85ajkU5cQLkDiEDZSxUPI57sWHBXQpuNOwanRHBaNKg==
    mode: 420
    size: 941
  index.js:
    integrity: sha512-GlwBoTIecaHCY/uDh4y7omhvcfVxtCwtQNk4WL8fTG/39iUSWmiTLCQ+derUqdLxjZsis/5Vw5/mYFpCsucGIA==
    mode: 420
    size: 3219
  package.json:
    integrity: sha512-/9Pz/tZde0iSFpFncsYYC1wXFN3eFufiyZ/1+KQhS7j4hKDB0GobGopLDWJSrYW1M7Fibw3Y1HTyKD9s39afbA==
    mode: 420
    size: 818
  test.js:
    integrity: sha512-9wIYD0Rlp3cSetaiETBt1WSZPBf309lAFgTfS8ZX2NiIkC3mMqTPVbVR3fufhLqKkAeIdz8Dx3g2FD75ThgsfA==
    mode: 420
    size: 1007
v3/files/5a/ed551de20b04af0a016254022499417f781a6384e39460ebfe77f1f2b08a5a14bb6d4a9dc1246063eaa1bda8499e66513ef7bcb1ab1d08cac3728c3f07c3ce-index.json:
  LICENSE:
    integrity: sha512-n8vl3Ia8M7208SVVnnZ8OqrZTHxLIbgLaU2bAVjuDxDuU7Enwr8F25dor9rx+2d4/QihlBL7PBra1lWGPmw1HQ==
    mode: 420
    size: 1088
  README.md:
    integrity: sha512-WAHlkulKpAdsF5GAlW8z+CYvpxV6Br/mNqaFS9O5st30R4D/oUBWQq2Ozi7WUU4ZooF4y+zc9WCKO+cQu/PA1A==
    mode: 420
    size: 5766
  index.js:
    integrity: sha512-F1S3Jxyg4w3H13WYZxGithspggDBL4Z73m9+GdVgR8mrd6kzBFIPoaMbDDpOb7azEPid2MtNlaXGYoB+9zP4Xw==
    mode: 420
    size: 662
  package.json:
    integrity: sha512-AdMGuA9LFnjGVfOwaF2x96dB/HRtJRwA6SUpiriWMKMWe96MfY13m3J3DEH58wK7Hj8xJnLOeqjmwVbDGe7q2g==
    mode: 420
    size: 1444

